driver/modern_bpf/add_new_syscalls.md

```diff
@@ -0,0 +1,530 @@
+
+1. Implement the BPF program following the `auxmap` template.
+2. Instruct our library `libpman` on how to call our BPF program implemented in step [1].
+3. Write a test to assert that the BPF program correctly works.
+
+#### 1. Implement the bpf program
+
+With the `auxamp` approach, the event dimension will be directly computed at run-time so we can start directly with the implementation!
+As for the `ringbuf` approach we have a template 👇
+
+```c
+/* if we are in the enter event */
+SEC("tp_btf/sys_enter") 
+int BPF_PROG(<your_syscall_name>_e,
+      struct pt_regs *regs,
+      long id)
+/* if we are in the exit event*/
+SEC("tp_btf/sys_exit")
+int BPF_PROG(<your_syscall_name>_x,
+      struct pt_regs *regs,
+      long ret)
+/**/
+{
+ struct auxiliary_map *auxmap = auxmap__get();
+ if(!auxmap)
+ {
+  return 0;
+ }
+
+ /* Fill the event header with your event type */
+ auxmap__preload_event_header(auxmap, <your_event>);
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ // The params we need to collect
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ /* We need to update the event header with the final event length, we know it only at this point */
+ auxmap__finalize_event_header(auxmap);
+
+ /* We push the event to userspace */
+ auxmap__submit_event(auxmap);
+
+ return 0;
+}
+```
+
+Let's fill in the template for our `access` syscall.
+
+```c
+SEC("tp_btf/sys_exit")
+int BPF_PROG(access_x,
+      struct pt_regs *regs,
+      long ret)
+{
+ struct auxiliary_map *auxmap = auxmap__get();
+ if(!auxmap)
+ {
+  return 0;
+ }
+
+ auxmap__preload_event_header(auxmap, PPME_SYSCALL_ACCESS_X);
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ /* As we saw before this is the comment for the first param */
+
+ /* Parameter 1: res (type: PT_ERRNO) */
+
+ /* Here we want to store the syscall return value, this param is a `PT_ERRNO`
+  * so as the doc says https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/helpers/store/auxmap_store_params.h#L258-L272
+  * we need to use the `auxmap__store_s64_param` helper
+  */
+ auxmap__store_s64_param(auxmap, ret);
+
+ /* Looking again at the `access` manual page https://man7.org/linux/man-pages/man2/access.2.html
+  * we can notice that `pathname` is the first param (so index `0`).
+  * This a pointer to the pathname so we use an `unsigned long` to store it.
+  * As the doc says https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/helpers/store/auxmap_store_params.h#L327-L340
+  * to store a `charbuf` param, so one between `PT_CHARBUF`, `PT_FSPATH`, `PT_FSRELPATH`, we need to use
+  * the `auxmap__store_charbuf_param` helper.
+  * - `MAX_PATH` is the maximum length we want to read from the pathname pointer
+  * - `USER` means that the provided pointer points to userspace memory and not kernel memory.
```

Comment:
Let me know if I should compile some on the kernel vs. user space memory :smile:

Reply:
Thanks... then let me write something up and share with you early next week (being on a training the rest of this week).